birdhouse/birdhouse-compose.sh

```diff
@@ -13,6 +13,34 @@
 # list of all variables to be substituted in templates
 #   some of these variables *could* employ provided values in 'default.env',
 #   but they must ultimately be defined one way or another for the server to work
+
+SHELL_EXEC_FLAGS=-e
+if [ x"${BIRDHOUSE_DEBUG_MODE}" = x"true" ]; then
+  SHELL_EXEC_FLAGS=-ex
+fi
+
+set ${SHELL_EXEC_FLAGS}
+
+exit_handler() {
+  exit_code="$?"
+  [ "$exit_code" -eq 0 ] && exit
+  [ "$BIRDHOUSE_EXPECTED_EXIT" = "true" ] && exit "$exit_code"
+  err_msg="An unexpected error occurred. See additional messages on stderr for details or rerun this command with BIRDHOUSE_DEBUG_MODE=true."
+  if command -v log >/dev/null; then
+    log ERROR "$err_msg"
+  else
+    >&2 echo "The following error occurred before logging could be initialized: $err_msg"
+  fi
+}
+
+expect_exit() {
+  exit_code="${1:-$?}"
+  export BIRDHOUSE_EXPECTED_EXIT=true
+  exit "${exit_code}"
+}
+
+trap 'exit_handler' EXIT
```

Comment:
Should this `exit_handler` and `expect_exit` be relocated to `read-configs.include.sh` so other scripts can also use it?

Reply:
I only worry that then we can't actually use it until read-configs.include.sh is sourced. But I agree that it's a useful tool to provide if other scripts want to use it. Let me see what I can do.

Reply:
The problem is that I want this to be defined early so that we get relevant error messages.

If we put it in read-configs.include.sh then we need to make sure that `COMPOSE_DIR` is defined first. In order to set `COMPOSE_DIR` there's a bunch of code that I'd like to be able to trap when it exits.

So I think I'd rather leave it here and other scripts can define their own error handling if needed. I just don't think it's worth making this less useful for birdhouse-compose.sh.

Reply:
I agree to leave it there to take advantage of error handling as early as possible.
Otherwise, we get situations similar to the logs were they cannot be called yet since the file is not sourced, and we have a few `echo <ERROR> <message>` mimicking its behavior in the meantime.